pkg/kapp/app/recorded_app.go

```diff
@@ -130,6 +130,77 @@ func (a *RecordedApp) UpdateUsedGVsAndGKs(gvs []schema.GroupVersion, gks []schem
 func (a *RecordedApp) CreateOrUpdate(labels map[string]string) error {
 	defer a.logger.DebugFunc("CreateOrUpdate").Finish()
 
+	foundMigratedApp, err := a.findAndUpdate(a.fqName(), labels)
+	if err != nil {
+		return err
+	}
+
+	if foundMigratedApp {
+		a.isMigrated = true
+		return nil
+	}
+
+	foundNonMigratedApp := false
+
+	if a.isMigrationEnabled() {
+		foundNonMigratedApp, err = a.findAndMigrate(a.name, a.fqName(), labels)
+	} else {
+		foundNonMigratedApp, err = a.findAndUpdate(a.name, labels)
+	}
+
+	if err != nil {
+		return err
+	}
+
+	if foundNonMigratedApp {
+		return nil
+	}
+
+	return a.create(labels)
+}
+
+func (a *RecordedApp) findAndUpdate(name string, labels map[string]string) (bool, error) {
+	cm, exists, err := a.find(name)
+
```

Comment:
nit: I do not think we need new lines after every call of `find()`
(I think this goes for all calls)

Reply:
Usually we check if the `err` is non nil and return if that's the case, but since I am checking for `exists` first, I thought of using a new line so that it's not confusing. But I can definitely remove the new line if it doesn't make sense to have one.

Reply:
I think it would have made sense if we were doing both checks later, but I think it makes sense to have no new line if we are checking values returned in the previous statement. (similar o what we do for `changed` or `isSecretCreated` in kctrl)
Once again, no strong opinions about this, just thought I would put it down

Reply:
Done.